nft_ingester/src/program_transformers/bubblegum/db.rs

```diff
@@ -441,3 +476,301 @@ where
 
     Ok(())
 }
+
+#[allow(clippy::too_many_arguments)]
+pub async fn upsert_asset_data<T>(
+    txn: &T,
+    id: Vec<u8>,
+    chain_data_mutability: ChainMutability,
+    chain_data: JsonValue,
+    metadata_url: String,
+    metadata_mutability: Mutability,
+    metadata: JsonValue,
+    slot_updated: i64,
+    reindex: Option<bool>,
+    raw_name: Vec<u8>,
+    raw_symbol: Vec<u8>,
+    seq: i64,
+) -> Result<(), IngesterError>
+where
+    T: ConnectionTrait + TransactionTrait,
+{
+    let model = asset_data::ActiveModel {
+        id: Set(id.clone()),
+        chain_data_mutability: Set(chain_data_mutability),
+        chain_data: Set(chain_data),
+        metadata_url: Set(metadata_url),
+        metadata_mutability: Set(metadata_mutability),
+        metadata: Set(metadata),
+        slot_updated: Set(slot_updated),
+        reindex: Set(reindex),
+        raw_name: Set(Some(raw_name)),
+        raw_symbol: Set(Some(raw_symbol)),
+        base_info_seq: Set(Some(seq)),
+        ..Default::default()
+    };
+
+    let mut query = asset_data::Entity::insert(model)
+        .on_conflict(
+            OnConflict::columns([asset_data::Column::Id])
+                .update_columns([
+                    asset_data::Column::ChainDataMutability,
+                    asset_data::Column::ChainData,
+                    asset_data::Column::MetadataUrl,
+                    asset_data::Column::MetadataMutability,
+                    // Don't update asset_data::Column::Metadata if it already exists.  Even if we
+                    // are indexing `update_metadata`` and there's a new URI, the new background
+                    // task will overwrite it.
+                    asset_data::Column::SlotUpdated,
+                    asset_data::Column::Reindex,
+                    asset_data::Column::RawName,
+                    asset_data::Column::RawSymbol,
+                    asset_data::Column::BaseInfoSeq,
+                ])
+                .to_owned(),
+        )
+        .build(DbBackend::Postgres);
+
+    // Do not overwrite changes that happened after decompression (asset_data.base_info_seq = 0).
+    // Do not overwrite changes from a later Bubblegum instruction.
+    query.sql = format!(
+        "{} WHERE (asset_data.base_info_seq != 0 AND excluded.base_info_seq >= asset_data.base_info_seq) OR asset_data.base_info_seq IS NULL",
+        query.sql
+    );
+    txn.execute(query)
+        .await
+        .map_err(|db_err| IngesterError::StorageWriteError(db_err.to_string()))?;
+
+    Ok(())
+}
+
+pub async fn lock_asset_creators_and_check_asset_base_info_seq<T>(
+    txn: &T,
+    id: Vec<u8>,
+    seq: i64,
+) -> Result<bool, IngesterError>
+where
+    T: ConnectionTrait + TransactionTrait,
+{
+    // Lock asset_creators table from any updates, but allow reads.
+    let lock_query = Statement::from_string(
+        DbBackend::Postgres,
+        "LOCK TABLE asset_creators IN EXCLUSIVE MODE".to_string(),
```

Comment:
Did you test this at scale? I'm concerned this will blow up indexing at high scale (e.g. over 500k cNFT mints coming in at once).

Reply:
Not needed in latest version which uses a no lock solution